### PR TITLE
Atomic ref helper

### PR DESCRIPTION
**Pull request by wenhuiNi**
Signed-off-by: Ni, Wenhui <wenhui.ni@intel.com>

### CODE REVIEW

File: clang/runtime/dpct-rt/include/atomic_ref_ext.hpp.inc [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+// AtomicRefExt|atomic_ref_ext
+// DPCT_DEPENDENCY_END
+// DPCT_CODE
+  /// atomically replaces the value of the referenced object with a non-atomic argument
+  /// \param [in]  replaces the value of the referenced object
+  /// \param operand The value to replace the pointed value.
+  /// \param memoryOrder The memory ordering used.
+  /// \param memoryScope The memory scope used.
+  void store(T operand, cl::sycl::memory_order memoryOrder = default_write_order,
+             cl::sycl::memory_scope memoryScope = default_scope) const noexcept {
```
Comment (ziranzha) [inline clang/runtime/dpct-rt/include/atomic_ref_ext.hpp.inc]: Is 'const' qualifier needed for store method?

File: clang/runtime/dpct-rt/include/atomic_ref_ext.hpp.inc [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+// DPCT_CODE
+  /// atomically obtains the value of the referenced object
+  /// \param [in, out]  replaces the value of the referenced object
+  /// \param memoryOrder The memory ordering used.
+  /// \param memoryScope The memory scope used.
+  /// \returns The value of the referenced object
+  T load(cl::sycl::memory_order memoryOrder = default_read_order,
+         cl::sycl::memory_scope memoryScope = default_scope) const noexcept {
+    cl::sycl::atomic_ref<T, DefaultOrder, DefaultScope, Space> atm(
+        *const_cast<T *>(&__d));
```
Comment (ziranzha) [inline clang/runtime/dpct-rt/include/atomic_ref_ext.hpp.inc]: How about use 'const_cast<T &>(__d)'?

File: clang/runtime/dpct-rt/include/atomic_ref_ext.hpp.inc [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+
+// DPCT_LABEL_BEGIN|atomic_ref_ext_1|dpct
+// DPCT_DEPENDENCY_BEGIN
+// AtomicRefExt|atomic_ref_ext
+// DPCT_DEPENDENCY_END
+// DPCT_CODE
+  /// no-argument default constructor
+  atomic_ref_ext() = default;
+  /// default constructor
+  atomic_ref_ext(T __d) : __d(__d){};
```
Comment (ziranzha) [inline clang/runtime/dpct-rt/include/atomic_ref_ext.hpp.inc]: How about use 'd' for the input arg name?

File: clang/runtime/dpct-rt/include/atomic_ref_ext.hpp.inc [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+// DPCT_LABEL_END
+
+// DPCT_LABEL_BEGIN|atomic_ref_ext_1|dpct
+// DPCT_DEPENDENCY_BEGIN
+// AtomicRefExt|atomic_ref_ext
+// DPCT_DEPENDENCY_END
+// DPCT_CODE
+  /// no-argument default constructor
+  atomic_ref_ext() = default;
+  /// default constructor
```
Comment (ziranzha) [inline clang/runtime/dpct-rt/include/atomic_ref_ext.hpp.inc]: Default constructor is constructor with no args.

File: clang/runtime/dpct-rt/include/atomic_ref_ext.hpp.inc (hunk 1 above)
Comment (wenhuiNi) [reply clang/runtime/dpct-rt/include/atomic_ref_ext.hpp.inc]: Yes, it is necessary to keep the keyword "const". The store function does not actually modify parameter __d in atomic_ ref_ext class, instead, modify what it points to. And in SYCL code,  "const' is also used. 

template <typename T, memory_order DefaultOrder, memory_scope DefaultScope,
          access::address_space AddressSpace>
class atomic_ref_base {
...

  void store(T operand, memory_order order = default_write_order,
             memory_scope scope = default_scope) const noexcept {
#ifdef __SYCL_DEVICE_ONLY__
    detail::spirv::AtomicStore(ptr, scope, order, operand);
#else
    (void)scope;
    ptr->store(operand, detail::getStdMemoryOrder(order));
#endif
  }

File: clang/runtime/dpct-rt/include/atomic_ref_ext.hpp.inc (hunk 2 above)
Comment (wenhuiNi) [reply clang/runtime/dpct-rt/include/atomic_ref_ext.hpp.inc]: fix it in commits 34e891d3e6db93f6f638490d0805d813e0bf7879

File: clang/runtime/dpct-rt/include/atomic_ref_ext.hpp.inc (hunk 3 above)
Comment (wenhuiNi) [reply clang/runtime/dpct-rt/include/atomic_ref_ext.hpp.inc]: fix it in commits 34e891d3e6db93f6f638490d0805d813e0bf7879

File: clang/runtime/dpct-rt/include/atomic_ref_ext.hpp.inc (hunk 4 above)
Comment (wenhuiNi) [reply clang/runtime/dpct-rt/include/atomic_ref_ext.hpp.inc]: fix it in commits 34e891d3e6db93f6f638490d0805d813e0bf7879

**Comment by wenhuiNi** [comment]
Conflicts with the version of the new merge. Close the PR and resubmit later

**Comment by wenhuiNi** [comment]
Conflicts with the version of the new merge. Close the PR and resubmit later

**Pull request closed without merging**